Bindings reordering fix

Crisspl opened the pull request.

devshgraphicsprogramming committed:
Let the work begin!

devshgraphicsprogramming committed:
Merge pull request #391 from buildaworldnet/master

Github this is not funny

devshgraphicsprogramming committed:
add Radeon Rays pre-2.0 as submodule, move OpenCL around a bit, update OpenCL and OpenGL headers

devshgraphicsprogramming committed:
Merge branch 'raytracing' of https://github.com/buildaworldnet/IrrlichtBAW into raytracing

devshgraphicsprogramming committed:
fix up modules pt.1

AnastaZIuk committed:
Merge pull request #7 from buildaworldnet/master

Important bug fixes

devshgraphicsprogramming committed:
Disabled GLM as a submodule because it annoys the fuck out of me (and added Radeon Rays)

devshgraphicsprogramming committed:
Radeon Rays builds and irr::ext::RadeonRays started

devshgraphicsprogramming committed:
update radeon rays submodule

devshgraphicsprogramming committed:
remove crappy stuff

AnastaZIuk committed:
Make CSTL and COBJ Loaders flipping work

AnastaZIuk committed:
Merge branch 'master' of https://github.com/buildaworldnet/IrrlichtBAW into newFlags2

pulling

AnastaZIuk committed:
Merge branch 'newFlags2' of https://github.com/AnastaZIuk/IrrlichtBAW into newFlags2

pulling

AnastaZIuk committed:
Fix memory leaks, make CX's flipping work

devshgraphicsprogramming committed:
add daily path tracer example

devshgraphicsprogramming committed:
big matrix touch up

devshgraphicsprogramming committed:
fix up examples 1-5 and Draw3DLine extension

devshgraphicsprogramming committed:
fix the determinant

devshgraphicsprogramming committed:
"fix-up the engine"

devshgraphicsprogramming committed:
fix hardware instancing for row-major matrix (and splines)

devshgraphicsprogramming committed:
fix instanced scene nodes pt. 2

devshgraphicsprogramming committed:
fix oit sample

devshgraphicsprogramming committed:
fix up remaining non-skinning samples (+ update Normal Cache)

devshgraphicsprogramming committed:
Fix Bullet Sample

devshgraphicsprogramming committed:
remove linux bin

devshgraphicsprogramming committed:
Finally fix Mitsuba camera loading

devshgraphicsprogramming committed:
fix original mitsuba loader

devshgraphicsprogramming committed:
Integrating RadeonRays is so easy

devshgraphicsprogramming committed:
work on deferred ray generation + set ID in radeon rays

devshgraphicsprogramming committed:
save progress

devshgraphicsprogramming committed:
add const mesh getters

devshgraphicsprogramming committed:
Touch up the API

devshgraphicsprogramming committed:
start the debug

devshgraphicsprogramming committed:
works but needs work

devshgraphicsprogramming committed:
fix minor bugs

devshgraphicsprogramming committed:
now time to skip the GPU-CPU-GPU roundtrip

AnastaZIuk committed:
Make CPLYMeshFileLoader work

AnastaZIuk committed:
Correct format #1

AnastaZIuk committed:
Correct format #2

AnastaZIuk committed:
Remove std::function, fix CXLoader bug

Matricies in CXLoader shouldn't be flipped

devshgraphicsprogramming committed:
Merge pull request #398 from AnastaZIuk/newFlags2

Make sure flipping work, debug and fix CX and CPLY loaders

devshgraphicsprogramming committed:
Fix reprojection

devshgraphicsprogramming committed:
update radeon rays used

devshgraphicsprogramming committed:
first half of GPU interop is off the ground

AnastaZIuk committed:
Add writerFlags as enum, make pseudo code

AnastaZIuk committed:
Fix semantic

devshgraphicsprogramming committed:
try some stuff

devshgraphicsprogramming committed:
wrong but closer

devshgraphicsprogramming committed:
accuracy fixes

devshgraphicsprogramming committed:
more accuracy fixes

devshgraphicsprogramming committed:
fix the depth reconstruction and raycasting

devshgraphicsprogramming committed:
Radeon Rays draws reflections

devshgraphicsprogramming committed:
environment light raytracing and mention of @manhnt9 in README.md

devshgraphicsprogramming committed:
THIS COMMIT BREAKS SKINNING

AnastaZIuk committed:
Create CSTLWriter flipping dependency

AnastaZIuk committed:
Make CSTLWriter work

AnastaZIuk committed:
Fix on bin CSTLWriting, forgot XD

AnastaZIuk committed:
Fix flags, make CPLYWriter work

AnastaZIuk committed:
Add missing header

devshgraphicsprogramming committed:
Merge pull request #401 from AnastaZIuk/writerFlags

Create WriterFlags and WriterFlipping

devshgraphicsprogramming committed:
Fix .x file loading crash and remove/fix a few screenshot invocations

devshgraphicsprogramming committed:
This is how you make a change to the BaW format and bump the version + add backwards compatibility.

devshgraphicsprogramming committed:
finish fixing skinning from .baw

devshgraphicsprogramming committed:
Prepare for stochastic light picking

devshgraphicsprogramming committed:
add some more utility functions

devshgraphicsprogramming committed:
get this many light sampling going!

devshgraphicsprogramming committed:
All shapes done except for cube

devshgraphicsprogramming committed:
all the math is probably wrong

devshgraphicsprogramming committed:
an attempt at cube rendering

devshgraphicsprogramming committed:
dumbsampling everything but cube and triangle is done

devshgraphicsprogramming committed:
cube "dumbsampling" done

devshgraphicsprogramming committed:
Triangle sampling done

devshgraphicsprogramming committed:
restore simple mitsuba scene to default

devshgraphicsprogramming committed:
better screenie

devshgraphicsprogramming committed:
adapt to all camera orientations

devshgraphicsprogramming committed:
introduce matrix - operator

devshgraphicsprogramming committed:
fix spectrum attribute handling in the mitsuba loader

devshgraphicsprogramming committed:
fix stupid bug

devshgraphicsprogramming committed:
add Sobol Sampler

devshgraphicsprogramming committed:
Turn sobol sampler into a Class

devshgraphicsprogramming committed:
Add owen scrambling

devshgraphicsprogramming committed:
Final Sampler Code.

devshgraphicsprogramming committed:
make sure API makes it clear Owen can be used with any LD sampler or sequence.

devshgraphicsprogramming committed:
Fix some Owen errors

devshgraphicsprogramming committed:
Raytracer now uses Sobol Sampler with Owen Scrambling.

TODO:
- Refactor Triangle Lights into a single light-importance entry/light
- Turn rectangle lights into triangle lights ?
- Split constant light into multiple triangular lights
- Apply Emissive on Lights
- Do self-light-occlusion
- Implement Light Visibility Importance Caching
- Improve Ellipsoid and Triangle sampling (try use direct methods)
- Turn disks into ellipsoids ?
- Turn cubes into rectangles/triangles ?

devshgraphicsprogramming committed:
test upper_bound

devshgraphicsprogramming committed:
remove test code

devshgraphicsprogramming committed:
Correct light CDF and PDF sampling and probability discrepancies

AnastaZIuk committed:
Fork GLM and GLEW, specify tags

devshgraphicsprogramming committed:
Merge pull request #413 from AnastaZIuk/glmGlewFix

Fork GLM and GLEW, specify tags

devshgraphicsprogramming committed:
I messed up cube and rectangle to triangle conversion.

devshgraphicsprogramming committed:
Correct some more CDF and PDF quantization mishaps

And convert constant area light to triangle approximation (will be useful later on for Importance Sampling according to Visibility*LuminareSolidAngle)

devshgraphicsprogramming committed:
Make lights visible

devshgraphicsprogramming committed:
stop the overoptimization that broke everything

devshgraphicsprogramming committed:
Fixed Constant Light to be triangles

devshgraphicsprogramming committed:
add a Mitsuba shape test.

Fix facingness of certain precomputed meshes

devshgraphicsprogramming committed:
Finally parity with Mitsuba in both left and right handed cameras.

Commit:
Added parallel-hashmap

devshgraphicsprogramming committed:
Merge branch 'raytracing' into master

devshgraphicsprogramming committed:
Merge pull request #417 from buildaworldnet/master

My own little merge from Master to RT

Commit:
Added phmap in Types.h

devshgraphicsprogramming committed:
All shapes finally fixed and backfaces too!

devshgraphicsprogramming committed:
Merge branch 'raytracing' of https://github.com/buildaworldnet/IrrlichtBAW into raytracing

devshgraphicsprogramming committed:
post merge fix-up

devshgraphicsprogramming committed:
Fix FOVAxis choice

devshgraphicsprogramming committed:
Rule of 5 patch

devshgraphicsprogramming committed:
lost instances temporary fix

devshgraphicsprogramming committed:
add a hash function to use with smart_refctd_ptr

devshgraphicsprogramming committed:
refactor mitsuba loader to make more sense

devshgraphicsprogramming committed:
add a gitignore for caches

devshgraphicsprogramming committed:
Merge pull request #419 from buildaworldnet/raytracing

Radeon Rays Raytracing Prototype

devshgraphicsprogramming committed:
Find Optix first attempt

devshgraphicsprogramming committed:
OptiX CMake done.

devshgraphicsprogramming committed:
Merge pull request #420 from buildaworldnet/optix_config

OptiX and CUDA CMake

Commit:
Compile fix

Commit:
Detect MSVC compiler

devshgraphicsprogramming committed:
Merge branch 'master' into parallel-hashmap

devshgraphicsprogramming committed:
Merge pull request #415 from WieszKto/parallel-hashmap

Added parallel-hashmap

Crisspl committed:
ahikhmin-shirley cos_eval finished with wrong shadowing term

Crisspl committed:
Merge pull request #22 from buildaworldnet/shader_pipeline

Sync up our shader branches [Mostly GLI]

Crisspl committed:
Fixing bindings reordering (done for graphics pipelines)

Basically done for graphics pipelines but stil dirty and untested

Crisspl committed:
Merge branch 'shaders' of https://github.com/Crisspl/IrrlichtBAW into shaders

Crisspl committed:
Forgot to push pipeline cache

Crisspl committed:
Fixed bindings reordering (tested)

TODO: a little API for pipeline caches + tests with it

devshgraphicsprogramming commented on include/IDriver.h:
```diff
@@ -293,6 +294,7 @@ class IDriver : public virtual core::IReferenceCounted, public IVideoCapabilityR
 		//! Create a renderpass independent graphics pipeline (@see ICPURenderpassIndependentPipeline)
         virtual core::smart_refctd_ptr<IGPURenderpassIndependentPipeline> createGPURenderpassIndependentPipeline(
+            IGPUPipelineCache* _pipelineCache,
             core::smart_refctd_ptr<IGPURenderpassIndependentPipeline>&& _parent,
```
we can get rid of `parent`

Crisspl replied:
at all? won't it be needed for vulkan implementation?

devshgraphicsprogramming replied:
nah, IHVs says that they don't care about parents at all.

Crisspl replied:
what are IHVs?

devshgraphicsprogramming replied:
independent hardware vendors

Crisspl committed:
Working with pipeline cache

Fixed (or at least made more legible) some code for layout compatibility
TODO: make GPU pipeline cache thread-safe
Keeping parsed spir-v in COpenGLPipelineCache turns out to be useless (i thought i had some idea before but i either lost it or it's really useless)

Crisspl committed:
COpenGLPipelineCache is now thread-safe

Crisspl committed:
Progress on CPU pipeline cache

Crisspl committed:
forgot to adjust this

devshgraphicsprogramming commented on include/IDriver.h:
```diff
@@ -42,6 +42,7 @@ class IDriver : public virtual core::IReferenceCounted, public IVideoCapabilityR
     protected:
 		core::smart_refctd_ptr<StreamingTransientDataBufferMT<> > defaultDownloadBuffer;
 		core::smart_refctd_ptr<StreamingTransientDataBufferMT<> > defaultUploadBuffer;
+        core::smart_refctd_ptr<IGPUPipelineCache> defaultPipelineCache;
```
don't make a default, let the user create one and pass it around.

devshgraphicsprogramming commented on include/irr/video/IGPUObjectFromAssetConverter.h:
```diff
@@ -576,7 +576,7 @@ inline created_gpu_object_array<asset::ICPURenderpassIndependentPipeline> IGPUOb
 
         (*res)[i] = m_driver->createGPURenderpassIndependentPipeline(
-            nullptr, //make and use some driver's default pipeline cache
+            m_driver->getDefaultPipelineCache(),
```
make it a parameter to the functions (also make a parameter struct to these functions of `IGPUObjectFromAssetConverter` for posteriority)

devshgraphicsprogramming commented on source/Irrlicht/CNullDriver.cpp:
```diff
@@ -134,11 +135,11 @@ bool CNullDriver::endScene()
-        const uint32_t lim = _destPplnLayouts[i] ? //if no descriptor set bound at this index
-            _destPplnLayouts[i]->isCompatibleUpToSet(IGPUPipelineLayout::DESCRIPTOR_SET_COUNT-1u, _newLayout) : 0u;
+        const int32_t lim = _destPplnLayouts[i] ? //if no descriptor set bound at this index
+            _destPplnLayouts[i]->isCompatibleUpToSet(IGPUPipelineLayout::DESCRIPTOR_SET_COUNT-1u, _newLayout) : -1;
```
doesnt this change affect a lot more?

Crisspl replied:
it does but I changed everything

devshgraphicsprogramming commented on src/irr/video/COpenGLPipelineCache.h:
```diff
@@ -48,13 +49,25 @@ class COpenGLPipelineCache final : public IGPUPipelineCache
 	}
 	//assumes that m_cache does not already contain an item with key==_key and layout fully compatible with _val.layout
-	void insert(SCacheKey&& _key, SCacheVal&& _val)
+	void insert(SCacheKey&& _key, SCacheVal&& _val, const asset::ICPUBuffer* _spirv)
```
you're caching the wrong thing here

Crisspl replied:
hm?

devshgraphicsprogramming replied:
we shouldn't be caching the SPIR-V shader, just the hash of the shader instead.

Crisspl replied:
yea, yea, ofc it does. This last param is just wrong approach to something else, doesnt matter, sorry. Will rework it today

devshgraphicsprogramming replied:
ok.

devshgraphicsprogramming commented on src/irr/video/COpenGLSpecializedShader.h:
```diff
@@ -43,6 +43,7 @@ class COpenGLSpecializedShader : public core::impl::ResolveAlignment<IGPUSpecial
 
 		const SInfo& getSpecializationInfo() const { return m_specInfo; }
 		const std::array<uint64_t, 4>& getSpirvHash() const { return m_spirvHash; }
+		const asset::ICPUBuffer* getSpirv() const { return m_spirv.get(); }
```
wrong thing to cache

devshgraphicsprogramming commented on src/irr/video/COpenGLSpecializedShader.cpp:
```diff
@@ -272,6 +272,7 @@ void COpenGLSpecializedShader::setUniformsImitatingPushConstants(const uint8_t*
 auto COpenGLSpecializedShader::compile(const COpenGLPipelineLayout* _layout) const -> std::pair<GLuint, SProgramBinary>
 {
 	spirv_cross::CompilerGLSL comp(reinterpret_cast<const uint32_t*>(m_spirv->getPointer()), m_spirv->getSize()/4ull);
```
can you cache this compiler?

devshgraphicsprogramming replied:
what did you end-up doing in the end?

Crisspl replied:
would u prefer it to be cached in COpenGLSPecShader or somehow in pipeline cache?

devshgraphicsprogramming replied:
pipeline cache, but in the second map/cache that's only for IGPUShader

devshgraphicsprogramming replied:
what happened in the end?

Crisspl replied:
COpenGLPipelineCache caches parsed spir-v and upon GL shader creation tries to find parsed spir-v to use it for faster spirv-cross compiler construction (in order to specialize, reorder bindings and generate final GLSL)

devshgraphicsprogramming commented on include/irr/video/IGPUPipelineCache.h:
```diff
@@ -13,6 +14,8 @@ class IGPUPipelineCache : public core::IReferenceCounted
 
 public:
 	virtual void merge(uint32_t _count, const IGPUPipelineCache** _srcCaches) = 0;
```
does Vulkan have this? If not then only CPU cache should have.

Crisspl replied:
yea, it has

Crisspl committed:
Got rid of pipeline parenting; extra parameters struct for cpu2gpu

devshgraphicsprogramming commented on src/irr/video/COpenGLSpecializedShader.h:
```diff
@@ -52,22 +55,21 @@ class COpenGLSpecializedShader : public core::impl::ResolveAlignment<IGPUSpecial
-		alignas(128) uint8_t m_uniformValues[IGPUMeshBuffer::MAX_PUSH_CONSTANT_BYTESIZE];
+		SInfo m_specInfo;
+		core::smart_refctd_ptr<const asset::ICPUBuffer> m_spirv;
+		std::array<uint64_t, 4> m_spirvHash;
```
shouldn't hash be in `COpenGLShader` the unspecialized one?

devshgraphicsprogramming replied:
also the SPIR-V?

Crisspl replied:
no, why? unspecialized shader has nothing to do with pipelines + unspecialized shader holds spirv OR glsl, while specialized always holds spirv

devshgraphicsprogramming replied:
> unspecialized shader holds spirv OR glsl

hash of either is fine.

Are there any mutations that get applied to the GLSL after we create an `IGPUShader` ?

Crisspl replied:
well, yea, like everything. #include resolvements, binding reorders, push constants to uniform translation, specialization constants resolvements

devshgraphicsprogramming replied:
> binding reorders, push constants to uniform translation, specialization constants resolvement

That's on the GLSL generated from SPIR-V.

But yeah I remember the run-time extension `include`s

devshgraphicsprogramming commented on src/irr/video/COpenGLSpecializedShader.h:
```diff
@@ -19,28 +20,30 @@ namespace video
-		{
-			if ((*m_GLnames)[_ctxID])
-				return (*m_GLnames)[_ctxID];
+		void setUniformsImitatingPushConstants(const uint8_t* _pcData, GLuint _GLname, const GLint* _locations, uint8_t* _state) const;
```
that can move to the pipeline since that will hold program objects now.

Crisspl replied:
not really because COpenGLSpecializedShader holds offsets of "uniforms"

devshgraphicsprogramming replied:
ok

Crisspl committed:
Fix build errors

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -2239,7 +2278,12 @@ void COpenGLDriver::SAuxContext::flushStateGraphics(uint32_t stateBits)
 			//assert(shdr); //this would be weird
 
-			shdr->setUniformsImitatingPushConstants(pushConstantsState[EPBP_GRAPHICS].data, this->ID);
+            //WARNING order of calls getShaderGLnameForCtx() and getUniformLocationsForStage() is important!!
```
why?

Crisspl replied:
because `getShaderGLnameForCtx` might create the GL name and `getUniformLocationsForStage` will need GL name to query locations (ofc locations are cached once queried, but order of those calls are still important for the first time when locations are yet unknown)

devshgraphicsprogramming replied:
> because getShaderGLnameForCtx might create the GL name

We could just create all the thread's programs (all the duplicates) at once during pipeline creation so we never, ever, stutter during a frame.

> getUniformLocationsForStage will need GL name to query locations (ofc locations are cached once queried, but order of those calls are still important for the first time when locations are yet unknown)

we should always compile/create at least one shader program upon pipeline creation and hence the uniform locations should be queried at that time and stored in the specialized shader (they will be identical for all following copies).

To avoid any hitches during rendering.

Crisspl replied:
done

Crisspl replied:
actually only the 2nd one
So u want to create for all threads at once? Now pipeline irrbaw objects store binaries and create GL names from it when a thread wants to use the pipeline

devshgraphicsprogramming replied:
there's probably an overhead in creating a new shader even from a binary, so maybe doing all up-front is a better idea.

devshgraphicsprogramming commented on src/irr/video/COpenGLRenderpassIndependentPipeline.h:
```diff
@@ -14,21 +14,35 @@ namespace video
         const asset::SPrimitiveAssemblyParams& _primAsmParams,
-        const asset::SRasterizationParams& _rasterParams
+        const asset::SRasterizationParams& _rasterParams,
+        uint32_t _ctxCount, uint32_t _ctxID, GLuint _GLnames[SHADER_STAGE_COUNT], COpenGLSpecializedShader::SProgramBinary _binaries[SHADER_STAGE_COUNT]
```
`const` the `_GLnames`

devshgraphicsprogramming commented on src/irr/video/COpenGLRenderpassIndependentPipeline.h:
```diff
@@ -53,9 +67,57 @@ class COpenGLRenderpassIndependentPipeline : public IGPURenderpassIndependentPip
+            (*m_uniformLocations[_stageIx])[i] = COpenGLExtensionHandler::extGlGetUniformLocation(GLname, uniforms_rng.begin()[i].m.name.c_str());
+
+        return m_uniformLocations[_stageIx]->data();
+    }
```
because you're using separable programs, the uniform locations can be precomputed (will be identical for all usages across all threads and pipelines) for a specialized shader (from a single compilation) and "cached" inside of the specialized shader

Crisspl replied:
they're precomputed once for all threads, however u're right that those can be stored in COpenGLSpecializedShader, i'll change it

devshgraphicsprogramming commented on src/irr/video/COpenGLPipelineCache.h:
```diff
@@ -0,0 +1,62 @@
+public:
+	struct SCacheVal {
+		COpenGLSpecializedShader::SProgramBinary binary;
+		core::smart_refctd_ptr<COpenGLPipelineLayout> layout;
```
shouldn't layout be the key? just like specialization would be?

Crisspl replied:
no because there's no way to make an operator< for it

devshgraphicsprogramming replied:
let's invent one (don't actually give it a `operator<` but use a `map` with a custom comparator functor struct.

You can treat a layout as a 4d vector/array of descriptor layouts, so implement a `<` operator like you would for a 4bit LSB int (first set `<`, then if first set `==` then check second set, and so on)

Then a `<` function for descriptor layouts could be defined the same as you define a `<` operator for an `std::array`

devshgraphicsprogramming replied:
its possible to implement a comparator using very similar logic to this:
https://github.com/buildaworldnet/IrrlichtBAW/pull/422/files#diff-73a7b1e23025a75063836a1f6c71770aR80

Crisspl replied:
so u're saying that having let's say layouts A, B,C and knowing relation of compatibility between them I can sort them in some order that will optimize process of searching the one compatible with some another layout D? I don't think so, how would it work? Compatibility between layouts A and B says nothing about compatibility A and D

Crisspl replied:
>its possible to implement a comparator using very similar logic to this:

u dont want just a comparator, u want a clearly defined relation

devshgraphicsprogramming replied:
> so u're saying that having let's say layouts A, B,C and knowing relation of compatibility between them I can sort them in some order that will optimize process of searching the one compatible with some another layout D? I don't think so, how would it work? 

Well in order for a SPIR-V to give you the same GLSL shader, the layout needs to be FULLY compatible, which means that all descriptor sets need to be identically defined.

Fun fact, how STL implements all algorithms in terms of just `<`:
A<B: `A<B`
A>B: `B<A`
A!=B: `A<B || B<A`
A==B: `!(A!=B)`

So we don't need a relation that makes sense, just one that is consistent (no `A<B<C && C<A` loop) and will guarantee that equality will work.

> Compatibility between layouts A and B says nothing about compatibility A and D
> u dont want just a comparator, u want a clearly defined relation

Well compatibility for a SHADER binding reordering requires full compatibility, so if A and B are compatible for shader X and C is compatible with A or B up to set 4, then C will be compatible as the descriptor set layouts are identically defined.

P.S. Another way to establish a `<` relation is to first compare shader hash, then specialization info, and then the computed (set,binding) to (type,binding) remappings which could be cached in `COpenGLPipelineLayout` similarily to precomputed multibind params 
https://github.com/Crisspl/IrrlichtBAW/blob/shaders/src/irr/video/COpenGLPipelineLayout.h#L73

devshgraphicsprogramming commented on src/irr/video/COpenGLPipelineCache.h:
```diff
@@ -0,0 +1,62 @@
+				return it->second.binary;
+		return {0,nullptr};
+	}
+	//assumes that m_cache does not already contain an item with key==_key and layout fully compatible with _val.layout
```
in `IRR_DEBUG` assert that

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -1323,14 +1323,45 @@ core::smart_refctd_ptr<IGPURenderpassIndependentPipeline> COpenGLDriver::createG
+        COpenGLPipelineLayout* layout = static_cast<COpenGLPipelineLayout*>(_layout.get());
+        COpenGLSpecializedShader* glshdr = static_cast<COpenGLSpecializedShader*>(_shader.get());
+
+        COpenGLPipelineCache::SCacheKey key{ glshdr->getSpirvHash(), glshdr->getSpecializationInfo() };
```
layout should be in the key though? since we want to get the full driver dependent shader binary (and that depends on GLSL generated from SPIR-V which depends on the layout)

Crisspl replied:
look in code that actually finds binaries

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -2282,8 +2313,11 @@ void COpenGLDriver::SAuxContext::flushStateGraphics(uint32_t stateBits)
-			shdr->setUniformsImitatingPushConstants(pushConstantsState[EPBP_GRAPHICS].data, GLname, locations, PCstate);
+			shdr->setUniformsImitatingPushConstants(pushConstantsState[EPBP_GRAPHICS].data, GLname, locations, PCstate, !uniformsEverSet);
+
+            currentState.pipeline.graphics.pipeline->afterUniformsSet(i, this->ID);
```
dont put this in the hot loop (stuttering) actually intialize at the start.

Crisspl replied:
i dont know what you mean actually

devshgraphicsprogramming replied:
what does `haveUniformsBeenEverSet` and `afterUniformsSet` do?

Crisspl replied:
former one gets flag and the latter one sets it to false

devshgraphicsprogramming replied:
can't the pipeline handle it in `setUniformsImitatingPushConstants` ?

Also why is `setUniformsImitatingPushConstants` a member of the shader!?

Crisspl replied:
because shader holds "uniforms' offsets"

devshgraphicsprogramming replied:
it can hold whatever it wants, but since pipelineholds both all the shaders, per-thread shader instances, and these actually hold state in driver's OpenGL implementation... the pipeline should be setting the uniforms with `setUniformsImitatingPushConstants` as a methdod not the shader

devshgraphicsprogramming replied:
?

devshgraphicsprogramming replied:
?

Crisspl replied:
done

devshgraphicsprogramming commented on source/Irrlicht/COpenGLDriver.cpp:
```diff
@@ -2393,7 +2427,12 @@ void COpenGLDriver::SAuxContext::flushStateCompute(uint32_t stateBits)
+            uint8_t* PCstate = currentState.pipeline.compute.pipeline->getPushConstantsStateForStage(0u, this->ID);
+            const bool uniformsEverSet = currentState.pipeline.compute.pipeline->haveUniformsBeenEverSet(0u, this->ID);
+
+			shdr->setUniformsImitatingPushConstants(pushConstantsState[EPBP_COMPUTE].data, GLname, locations, PCstate, !uniformsEverSet);
```
same as https://github.com/buildaworldnet/IrrlichtBAW/pull/422/commits/98fcf894ea6ffb42cdbe37ba98615dab9fe4e6f8#r379923220

devshgraphicsprogramming commented on src/irr/video/COpenGLPipelineCache.h:
```diff
@@ -33,19 +34,31 @@ class COpenGLPipelineCache final : public IGPUPipelineCache
-			m_cache.insert(src.begin(), src.end());
+			src.outputAll(sz, buf->data());
+			for (size_t j = 0ull; j < sz; ++j)
+				m_cache.insert((*buf)[j].first, (*buf)[j].second);
```
use a single-threaded cache, and mutex the shit out of the pipeline, just using a thread-safe cache won't make the member functions of the owner safe to call from multiple threads.

Crisspl replied:
> just using a thread-safe cache won't make the member functions of the owner safe to call from multiple threads.

why?

Crisspl replied:
hm, ok, i know what u mean, nvm

Crisspl committed:
Moved uniform location caching to COpenGLSPecializedShader:

devshgraphicsprogramming commented on src/irr/video/COpenGLSpecializedShader.h:
```diff
@@ -61,6 +61,7 @@ class COpenGLSpecializedShader : public core::impl::ResolveAlignment<IGPUSpecial
 		//mutable bool m_uniformsSetForTheVeryFirstTime = true;
 		//alignas(128) uint8_t m_uniformValues[IGPUMeshBuffer::MAX_PUSH_CONSTANT_BYTESIZE];
 		core::vector<SUniform> m_uniformsList;
+		mutable core::smart_refctd_dynamic_array<GLint> m_locations;
```
either use `vector` for both or `smart_Refctd_dynamic_array`

Crisspl committed:
COpenGLDriver always creates one GL name per active stage upon pipeline creation and queries uniform locations

devshgraphicsprogramming commented on src/irr/video/COpenGLSpecializedShader.cpp:
```diff
@@ -291,6 +288,9 @@ auto COpenGLSpecializedShader::compile(const COpenGLPipelineLayout* _layout) con
 	if (logbuf[0])
 		os::Printer::log(logbuf, ELL_ERROR);
 
+	if (m_locations.empty())
```
what if there are no uniforms at all? will we keep on hitting this?

devshgraphicsprogramming replied:
ah nvm its in `compile` so only one hit ever.

Crisspl replied:
not ever, every time pipeline with this shader is created. Yes, we hit every time then. And then it runs 0 iterations loop so whatever

devshgraphicsprogramming replied:
yeah, I realized

devshgraphicsprogramming commented on src/irr/video/IOpenGLPipeline.h:
```diff
@@ -0,0 +1,91 @@
+            if (n != 0u)
+                COpenGLExtensionHandler::extGlDeleteProgram(n);
+        _IRR_ALIGNED_FREE(m_uniformValues);
+    }
```
you're leaking binaries?

Crisspl replied:
no, binaries are stored in smart array. It's in a raw one because of alignment

devshgraphicsprogramming commented on src/irr/video/IOpenGLPipeline.h:
```diff
@@ -0,0 +1,91 @@
+    {
+        auto uniforms_rng = _shader->getUniforms();
+
+        auto uniformLocations = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<GLint>>(uniforms_rng.length());
```
do we always need to make a new array?

Crisspl replied:
not always, this code runs just once. Anyway it's no longer there, now locations are stored in COpenGLSpecializedShader

devshgraphicsprogramming commented on src/irr/video/IOpenGLPipeline.h:
```diff
@@ -0,0 +1,91 @@
+    //mutable for deferred GL objects creation
+    mutable core::smart_refctd_dynamic_array<GLuint> m_GLnames;
+    mutable core::smart_refctd_dynamic_array<COpenGLSpecializedShader::SProgramBinary> m_shaderBinaries;
+    mutable core::smart_refctd_dynamic_array<bool> m_uniformsSetForTheVeryFirstTime;
```
make these `std::array`, there's no need for dynamic allocation for arrays we 100% know the upper bound for.

Crisspl replied:
so what's the upper bound?

devshgraphicsprogramming replied:
argh shit `m_GLnames` and `m_uniformsSetForTheVeryFirstTime` actually needs a dynamic allocation (but you could join the `GLuint` and `bool` into a single struct/pair and only have one dynamically allocated array).

But `m_shaderBinaries` you only need up to MAX_STAGES_PER_PIPELINE (which is your template paramater right now).

Crisspl committed:
COpenGL*Pipeline creates shaders for all threads upfront

Crisspl committed:
COpenGLPipelineCache is actually thread-safe

Crisspl committed:
almost merged with devsh/master

TODO: CSTLMeshFileLoader.h,.cpp and TypedBlob.cpp

Crisspl committed:
tmp cmake fix

Crisspl committed:
Finished merge with devsh/master

devshgraphicsprogramming commented on CMakeLists.txt:
```diff
@@ -23,17 +23,17 @@ project(Irrlicht LANGUAGES CXX C)
 # CUDA test
 include(CheckLanguage)
 check_language(CUDA)
-enable_language(CUDA)
```
what error did you get withthis?

devshgraphicsprogramming replied:
https://cmake.org/cmake/help/v3.2/module/CheckLanguage.html

hmm seems I should check for `CMAKE_CUDA_COMPILER` before proceeding to enable language

Crisspl replied:
>what error did you get withthis?

"no CUDA toolset found"

Crisspl replied:
ok i fixed it

Crisspl committed:
Merge branch 'shader_pipeline' of github.com:buildaworldnet/IrrlichtBAW into shaders

Crisspl committed:
Merge branch 'shader_pipeline' of github.com:buildaworldnet/IrrlichtBAW into shaders

Crisspl committed:
Updated ex03 regarding AMD driver bug

Crisspl committed:
Added parsed spirv cache to COpenGLPipelineCache

devshgraphicsprogramming commented on src/irr/video/COpenGLPipelineCache.h:
```diff
@@ -33,52 +33,75 @@ class COpenGLPipelineCache final : public IGPUPipelineCache
 	}
-	//assumes that m_cache does not already contain an item with key==_key and layout fully compatible with _val.layout
-	void insert(SCacheKey&& _key, SCacheVal&& _val, const asset::ICPUBuffer* _spirv)
+	const spirv_cross::ParsedIR* findParsedSpirv(const std::array<uint64_t, 4>& _key)
```
maybe make the argument an `ICPUShader` then the `ICPUShader` can keep track of its own hash?

Crisspl replied:
why? pipeline operates on specialized shaders so why even bother with unspecialized ones?

devshgraphicsprogramming replied:
but you're findind the parsed spirv of the unspecialized shader right?

Crisspl replied:
theres no spirv of unspecialized shader and spirv of specialized shader. Theres just spirv. Specialized shader is spirv + spec info

devshgraphicsprogramming replied:
well we can compute the hash of spir-v and GLSL the same anyway, and make ICPUShader keep that.

devshgraphicsprogramming commented on src/irr/video/COpenGLPipelineCache.h:
```diff
@@ -33,52 +33,75 @@ class COpenGLPipelineCache final : public IGPUPipelineCache
+	{
+		const std::lock_guard<std::mutex> _(m_bin_cache_mutex);
 #ifdef _IRR_DEBUG
 		assert(!find(_key, _val.layout.get()).binary);
```
wrong assert, should be `!found || !isCompatible(found.layout.get(),_val.layout.get()`

Crisspl replied:
see code of `find()`

devshgraphicsprogramming replied:
`find` assumes that `layout` is compatible with the key?

Crisspl replied:
`find` doesnt find anything (returns nullptr) if passed layout is not compatible with any of entries of equal key

Crisspl committed:
Fixed cmake regarding CUDA

devshgraphicsprogramming committed:
Make an example displaying a cube using a geometry creator

devshgraphicsprogramming committed:
add missing file

Crisspl committed:
COpenGLExtensionHandler: added bindless textures GL calls

Crisspl committed:
forgot about definitions

devshgraphicsprogramming committed:
Make and test an example of using geometry creator (cone and arrow left for Wiesz_Kto)

It has to be tested on  OpenGL state caching side, something went wrong

devshgraphicsprogramming committed:
Merge branch 'shaders' of https://github.com/Crisspl/IrrlichtBAW into geometryCreator

devshgraphicsprogramming committed:
add missing file

Crisspl committed:
Merge branch 'geometryCreator' of github.com:AnastaZIuk/IrrlichtBAW into shaders

Crisspl committed:
Fixed bug in COpenGLDriver regarding sometimes no rebinding of vertex and index buffers

Crisspl committed:
AMD driver bug fix/workaround

as for now takes into account only main() function

AnastaZIuk committed:
Prepare CPLY for refactoring

AnastaZIuk committed:
Refactor CPLY writer

Crisspl committed:
Further semi-fix of AMD driver bug

AnastaZIuk committed:
Change handling creating mesh buffer function, adjust code to devsh's comments

Crisspl committed:
Moved setUniformsImitatingPushConstants() to pipeline class

Crisspl committed:
COpenGLPipelineCache: merged layout into cache key

Crisspl committed:
amd fix: should now work for whole shader (not just entry point func)

devshgraphicsprogramming commented on src/irr/video/COpenGLPipelineCache.cpp:
```diff
@@ -3,6 +3,76 @@
+	if (lendiff != 0)
+		return lendiff;
+
+	const auto* lhs = A->getBindings().begin();
+	const auto* rhs = B->getBindings().begin();
```
I'd compare the OpenGL multibind arguments here instead, cause after all you only care if the bindings got remapped the same to OpenGL slots (which means that as you iterate through the `(set,binding)` tuples of both layouts, do all 4 running total counts of UBO, SSBO, Texture and Image stay the same?)

But I guess that can be left as a TODO in the comment, since this seems to be "good enough"

AnastaZIuk committed:
Fix flipping and bitmask, insert shaders into cache

AnastaZIuk committed:
Use default buildin Assets (pipeline layout, shaders)

AnastaZIuk committed:
Perform tests, indices are funcked up and don't know really why

Crisspl committed:
Merge pull request #26 from AnastaZIuk/CPLY

CPLY Loader and Writer

AnastaZIuk committed:
pushing for catching my mistakes

Crisspl committed:
Merge branch 'CPLY' of github.com:AnastaZIuk/IrrlichtBAW into shaders

Crisspl committed:
some help with PLY loader

Crisspl committed:
Merge branch 'shaders' of https://github.com/Crisspl/IrrlichtBAW into shaders

devshgraphicsprogramming commented on src/irr/asset/CPLYMeshFileLoader.cpp:
```diff
@@ -643,7 +643,7 @@ bool CPLYMeshFileLoader::genVertBuffersForMBuffer(asset::ICPUMeshBuffer* _mbuf,
 		inputParams.enabledBindingFlags |= currentBitmask;
 		inputParams.enabledAttribFlags |= currentBitmask;
-		inputParams.bindings[attrib] = { asset::getTexelOrBlockBytesize((E_FORMAT)vertexAttribParamsAllOptions[attrib].format), EVIR_PER_INSTANCE };
+		inputParams.bindings[attrib] = { 16, EVIR_PER_VERTEX };
```
cmon guys, you know I hate constant literals.
